orderer/common/cluster/util_test.go

```diff
@@ -1187,3 +1189,41 @@ qz8bY+s6AiBvO0aOfE8M4ibjmRE4vSXo0+gkOIJKqZcmiRdnJSr8Xw==
 	expected := `{"CAs":[{"Expired":false,"Issuer":"self","Subject":"SERIALNUMBER=65764076798709079104397911007043976879"},{"Expired":true,"Issuer":"SERIALNUMBER=65764076798709079104397911007043976879","Subject":"SERIALNUMBER=187004237184026939685179573732217690720"}],"Endpoint":"orderer.example.com:7050"}`
 	require.Equal(t, expected, actual)
 }
+
+func TestComparisonMemoizer(t *testing.T) {
+	var invocations int
+
+	m := &cluster.ComparisonMemoizer{
+		MaxEntries: 5,
+		F: func(a, b []byte) bool {
+			invocations++
+			return bytes.Equal(a, b)
+		},
+	}
+
+	// Warm-up cache
+	for i := 0; i < 5; i++ {
+		notSame := m.Compare([]byte{byte(i)}, []byte{1, 2, 3})
+		assert.False(t, notSame)
```

Comment:
Mostly a nit, but most all of these asserts would probably be better as requires.

Reply:
I don't see why? A require is needed in case the test might fail further along in the flow, which is like when you check something is not nil and then a require would result in a nil pointer panic. An assert is perfectly fine here, especially since the test is small and serial and easy to debug. I changed to assert nonetheless, just to adhere to our (false) coding bible.

Reply:
My assertion would be that any time a failure higher in a test case would cause future failures to be expected, you are better to require than to assert.

For example, testing how a sqrt function handles positive, negative, and zero cases would make sense as asserts

```
assert.Equal(t, mySqrt(9.0), 3.0)
assert.True(t, math.IsNan(mySqrt(-9.0)))
assert.Equal(t, mySqrt(0), 0)
```

Here, any single case failing has no impact on the future cases, and, seeing all three failures would be useful when trying to address any bugs.

On the other hand, the test you have here, if the cache inappropriately fails to handle insertion of the initial certificates, then the future failures are meaningless and expected.  Why continue executing and producing additional failure output when it provides negative value?

Reply:
> On the other hand, the test you have here, if the cache inappropriately fails to handle insertion of the initial certificates, then the future failures are meaningless and expected. 

I see what you mean in general, but we don't know that from the test. The only thing we know is that it returned an unexpected outcome. It may cache the value correctly despite returning the wrong result.

> Why continue executing and producing additional failure output when it provides negative value?

You're right in theory but I'm not sure it matters that much in practice. We've been using assert all over the project and now we moved almost all of them to require and I haven't noticed any difference, but maybe that's just me.